### ISSUE TITLE
Order Object Permissions in the Group Editor

### ISSUE THREAD

**Issue by dkirkham** [wagtail/wagtail#10468]
### Is your proposal related to a problem?

The Group Editor - at menu->Settings->Groups or /admin/groups/<<int:group>>/ - allows an administrator to control the permissions each member of the group receives. The "Object Permissions" section, which allows add, change, delete and other model permissions to be edited, are listed in an unmanaged order. The order appears to depend on the migration order, and depending on the development and deployment workflow, results in that ordering being different between development, staging and production deployments.

### Describe the solution you'd like

I would like a solution where the developer can control the order in which each Object type appears. Object permissions can then be placed in a logical order, such as placing related objects together. That order will then be consistent between deployments.

The solution should address objects defined by the developer and within Wagtail itself.

### Describe alternatives you've considered

There appears to be no practical solution to this problem.

### Additional context

I've developed a PR that addresses this, which I'll post shortly.

**Comment by dkirkham** [comment]
Hi @laymonage, the way I would prefer to order the object permissions is in sets of related objects, where the same permissions would typically be applied. Within each object set, the order may be alphabetical or it may be in some other order, eg. relating to the object hierarchy, importance, etc. It is for this reason that I feel a simple alphabetical sort is only a minor improvement.

**Comment by dkirkham** [comment]
Hi @lb-, regarding implementing a default ordering for objects listed in the user group editing, I think it is best discussed and settled here, and then I'll look at how that impacts #10469. I agree that whatever is done on this issue, it should not affect how snippet registration is handled.

I think the best option for a default is an alphabetical ordering which includes the app_label and the model name. This will ensure the order is essentially stable, and will by and large place related models next to each other in the list. However, I note that the app_label isn't shown in the object list, so some may prefer the alphabetical ordering to be based on just the model name.

The next question is how this default ordering interacts with a developer-defined ordering. In my PR, the code places object types that weren't given a defined ordering _after_ those which were. I could continue this approach, which again is stable whether or not the developer fully defines an order, or leaves some or all of the objects to the default order. To be honest, I can't think of an alternative approach that would be sensible and straightforward to document. Suggestions welcome!

**Comment by dkirkham** [comment]
As I commented in the PR, there is already a default ordering for the Permission class within the Django source code, which is the ordering I proposed above - by app_label and the model name. The revised PR will preserve this order in the absence of any specifically configured object ordering.

I have retained the behaviour where any specifically ordered objects will precede all default ordered objects in the group editor list.
